### ISSUE TITLE
Support ethr DID method in the JS IDM Wallet

### ISSUE THREAD

**Issue by satazor** [ipfs-shipyard/pm-idm#252]
## Description

Support [ethr](
https://github.com/uport-project/ethr-did) DID method in the JS IDM Wallet.

For now, the `idm-wallet` module just supports the `IPID` did method. Though, we should add at least one more to prove its concept. I would suggest adding `ethr`, which is the latest DID method used by uport (I think).

To do this, we just need to add `ethr.js` and change [`didm/index.js`](https://github.com/ipfs-shipyard/js-idm-wallet/tree/master/src/didm) to forward all `ethr` did methods to its correspondent functions.

## Acceptance Criteria

- [ ] Implement `ethr.js` which uses https://github.com/uport-project/ethr-did
- [ ] The `ethr.js` should have the exact same interface as the existent `ipd.js` file
- [ ] The key generation from the mnemonic should be the same that uPort uses, which is bip32. I think we can use the already existent `human-crypto-keys`, but needs to be tested.

**Comment by satazor** [comment]
@vasa-develop

- The interface of all methods in `ethr.js` should be exactly the same as `ipid, meaning that the adapter layer will be a little bit more complicated.

As an example: the `operations` parameter of both `create()` and `update()` is a function that is called to mutate the document and is being used by the `identities` scope, see: https://github.com/ipfs-shipyard/js-idm-wallet/blob/ce90f54ea1b97c9af2e1a9f6760d53daf1e59e43/src/identities/index.js#L110. The `operations` function (as well as many others) is not yet documented in the spec, but you may find documentation here: https://github.com/ipfs-shipyard/js-did-ipid#operations. In that adapter layer, you would need to map any functions the `document` parameter of the `operation` function has to the `setAttribute` of `ethr.js`.

-  Consider lazy loading `ethr.js` with dynamic import so that it doesn't weight much in the bundle size. We can do this in a separate issue, we just need to create the issue so that we won't forget.

**Comment by satazor** [comment]
@oed hello! Quick question: is the `ethr` the most recent did method used by uport?

**Comment by oed** [comment]
it is yes :)

**Comment by vasa-develop** [comment]
UPDATE:
ok. Finally, I have all the components for `did-ethr` working.
They really do have crappy docs and the incompatible web3 provider makes it even worse.

**Comment by vasa-develop** [comment]
@satazor 

>  In that adapter layer, you would need to map any functions calls to the document parameter (of the operation function) to the setAttribute of ethr-did.

Here, is the file https://github.com/ipfs-shipyard/js-idm-wallet/blob/ce90f54ea1b97c9af2e1a9f6760d53daf1e59e43/src/identities/index.js the adapter layer? 
I am asking this for figuring out where should I add the code for mapping the `setAttribute` function.

**Comment by satazor** [comment]
The adapter layer is the `didm` scope, see https://github.com/ipfs-shipyard/js-idm-wallet/blob/master/src/didm/index.js

**Comment by vasa-develop** [comment]
@satazor 
[Here](https://github.com/ipfs-shipyard/js-idm-wallet/blob/ce90f54ea1b97c9af2e1a9f6760d53daf1e59e43/src/didm/methods/ipid.js#L38), in this line we execute `addPublicKey` on the `document` object. 
But then we also execute the same `addPublicKey` on the `document` object in the [`identities` scope](https://github.com/ipfs-shipyard/js-idm-wallet/blob/ce90f54ea1b97c9af2e1a9f6760d53daf1e59e43/src/identities/index.js#L110) in the `options` callback.
So, basically, the `addPublicKey` is executed on the `document` object.
Can you explain it a bit to me?

**Comment by vasa-develop** [comment]
@satazor 
For now, I am using ethereum specific libs to regenerate the keyPair or private key from the `param`.
But going forward, I think we should add the support for it with the `human-crypto-keys` package.

Does that sound ok to you?

**Comment by vasa-develop** [comment]
@satazor 
Why we don't add the [`createDidIpid`](https://github.com/ipfs-shipyard/js-idm-wallet/blob/master/src/didm/methods/ipid.js#L72) method in the [`constructor`](https://github.com/ipfs-shipyard/js-idm-wallet/blob/master/src/didm/methods/ipid.js#L16) itself?

**Comment by satazor** [comment]
> @satazor
> For now, I am using ethereum specific libs to regenerate the keyPair or private key from the `param`.
> But going forward, I think we should add the support for it with the `human-crypto-keys` package.
> 
> Does that sound ok to you?

Yep. What is it missing in `human-crypto-keys` to support key pair generation for `ethr`?


> @satazor
> Why we don't add the [`createDidIpid`](https://github.com/ipfs-shipyard/js-idm-wallet/blob/master/src/didm/methods/ipid.js#L72) method in the [`constructor`](https://github.com/ipfs-shipyard/js-idm-wallet/blob/master/src/didm/methods/ipid.js#L16) itself?

Unfortunately we can’t because it’s async, and constructors in JS can’t be async.

**Comment by vasa-develop** [comment]
> Yep. What is it missing in human-crypto-keys to support key pair generation for ethr?
I need to have `secp256k1` algo support for key generation.
Plus, I need to have the key exported in `hex` format. So, also support for that.

P.S. Do I have write permission for this repo? I am not able to push to a new branch in this repo.
`

**Comment by vasa-develop** [comment]
@satazor 
There seems to be no definition of a device key in the `ethr-did` [specification](https://github.com/uport-project/ethr-did-resolver/blob/develop/doc/did-method-spec.md).

So, how do we add a device key? Any thoughts?

There are definitions for [delegate keys](https://github.com/uport-project/ethr-did-resolver/blob/develop/doc/did-method-spec.md#delegate-keys) which can act as signing keys.

**Comment by vasa-develop** [comment]
Are there any weekly/bi-weekly calls for IDM?
If yes, then please add me to the list. It will be better to discuss these issues over a call so that we can discuss probable problems and their solutions.

**Comment by vasa-develop** [comment]
@satazor 
The [identity](https://github.com/ipfs-shipyard/js-idm-wallet/tree/master/src/identities/identity) folder makes heavy use of orbitDB for its working.
Should I have to create method-specific `identity` folders, that can be imported in the `identities` [index.js](https://github.com/ipfs-shipyard/js-idm-wallet/blob/master/src/identities/index.js), and managed from there?

**Comment by satazor** [comment]
@vasa-develop Yes we have bi-weekly progress calls, but we are figuring out the roadmap for the next quarters. They are in standby for now and we hope to resume them very soon.

**Comment by satazor** [comment]
@vasa-develop in terms of replication provider per identity. This would be really hard to do, since each DID-method is doing it's own thing. Later on, having a data provider per identity makes sense but it's a beast that needs to be tamed at his own time.

**Comment by vasa-develop** [comment]
> Later on, having a data provider per identity makes sense but it's a beast that needs to be tamed at his own time.

Seems like it deserves a separate issue.

**Comment by vasa-develop** [comment]
@satazor 
What about [this issue](https://github.com/ipfs-shipyard/pm-idm/issues/252#issuecomment-514290919)?
